### PR TITLE
Feature/serializer

### PR DESCRIPTION
**Pull request by Tehsapper**
implemented material serializer & deserializer

### CODE REVIEW

File: include/core/uniform.hpp [hunk 1]
```diff
@@ -5,21 +5,40 @@
...
+	std::unique_ptr<Uniform> deserializeUniformValue(Reader&& reader);
 
     class Uniform {
     protected:
         std::string name;
         UniformType type;
         UniformValueType value_type;
         bool changed;
+
+        virtual ByteWriter& serialize(ByteWriter& writer) const {
```
Comment (hotstreams) [inline include/core/uniform.hpp:34]: make better serialization interface
move it to cpp file

File: include/core/uniform.hpp [hunk 2]
```diff
@@ -30,16 +49,25 @@ namespace GraphicsEngine {
 
         [[nodiscard]] virtual Uniform* clone() noexcept = 0;
         virtual void set(const ShaderProgram& shader) = 0;
+
+        friend ByteWriter& operator<<(ByteWriter& writer, const Uniform& u) {
```
Comment (hotstreams) [inline include/core/uniform.hpp:53]: separate definition from declaration

File: include/core/uniform.hpp [hunk 3]
```diff
@@ -30,16 +49,25 @@ namespace GraphicsEngine {
...
     template<typename T>
     class UniformValue : public Uniform {
     protected:
         T value;
 
-        constexpr UniformValueType getUniformValueType();
+        [[nodiscard]] constexpr UniformValueType getUniformValueType() const;
         UniformValue(const std::string& name, UniformType type, const T& value);
+
+        ByteWriter& serializeFields(ByteWriter& writer) const;
```
Comment (hotstreams) [inline include/core/uniform.hpp:66]: consider using more template/constexpr/compile-time language features

File: include/core/uniform.hpp [hunk 4]
```diff
@@ -54,7 +82,11 @@ namespace GraphicsEngine {
     private:
         std::shared_ptr<Texture> sampler;
         GLuint sampler_id {};
+    protected:
+    	ByteWriter& serialize(ByteWriter& writer) const override;
     public:
+        explicit UniformSampler(ByteReader& reader);
```
Comment (hotstreams) [inline include/core/uniform.hpp:88]: do not want

File: include/core/uniform.hpp [hunk 5]
```diff
@@ -5,21 +5,40 @@
 #include <chrono>
 #include <core/texture_visitor.hpp>
 #include <core/context_debug.hpp>
+#include <util/bytewriter.hpp>
```
Comment (hotstreams) [inline include/core/uniform.hpp:8]: do not want

File: include/material_system/material.hpp [hunk 6]
```diff
@@ -59,6 +64,12 @@ namespace GraphicsEngine {
 
         [[nodiscard]] virtual Material* clone() const noexcept;
 
+        static constexpr uint32_t serial_version = 0x1;
```
Comment (hotstreams) [inline include/material_system/material.hpp:67]: h3h

File: include/material_system/material.hpp [hunk 7]
```diff
@@ -59,6 +64,12 @@ namespace GraphicsEngine {
...
+
+	    void serialize(ByteWriter& writer);
+
+	    static std::shared_ptr<Material> deserialize(ByteReader& reader);
```
Comment (hotstreams) [inline include/material_system/material.hpp:71]: why not materialserializer class method?

File: include/material_system/material_builder.hpp [hunk 8]
```diff
@@ -30,6 +30,7 @@ namespace GraphicsEngine {
 
         MaterialBuilder& add(PropertyType type, std::shared_ptr<Texture> texture);
         MaterialBuilder& add(PropertyType type, const glm::vec4& value);
+        MaterialBuilder& add(PropertyType type, std::unique_ptr<Uniform> uniform);
```
Comment (hotstreams) [inline include/material_system/material_builder.hpp:33]: nice property ruin

File: src/core/uniform.cpp [hunk 9]
```diff
@@ -126,6 +145,29 @@ void UniformSampler::set(const ShaderProgram& shader) {
     return new UniformSampler(*this);
 }
 
+ByteWriter& UniformSampler::serialize(ByteWriter& writer) const {
+	if (sampler->getName()) {
+		writer << UniformType::Sampler;
+		UniformValue<int>::serializeFields(writer);
```
Comment (hotstreams) [inline src/core/uniform.cpp:151]: you do not need serialize current texture unit GPU bind : /

File: src/core/uniform.cpp [hunk 10]
```diff
@@ -126,6 +145,29 @@ void UniformSampler::set(const ShaderProgram& shader) {
...
+	}
+}
+
+UniformSampler::UniformSampler(ByteReader& reader) : UniformValue{reader} {
+	std::string path;
+	reader >> path;
+	try {
+		sampler = assets.textures.at(path);
+	}
+	catch (const resource_container_error& err) {
```
Comment (hotstreams) [inline src/core/uniform.cpp:164]: why not ... or resource_container_error ?

File: src/core/uniform.cpp [hunk 11]
```diff
@@ -221,3 +263,30 @@ namespace GraphicsEngine {
     template class UniformValue<glm::mat3>;
     template class UniformValue<glm::mat4>;
 }
+
+std::unique_ptr<Uniform> GraphicsEngine::deserializeUniformValue(ByteReader& reader) {
```
Comment (hotstreams) [inline src/core/uniform.cpp:267]: want copy elission

File: src/material_system/material.cpp [hunk 12]
```diff
@@ -213,6 +218,42 @@ Material::Material() noexcept
...
+		    .create(reader)
+		    .setBlending(reader)
+		    .setShading(reader);
+
+		{
+			size_t size {reader};
+			for (size_t i = 0; i < size; ++i) {
+				PropertyType ptype {reader};
+				auto uptr = deserializeUniform(reader);
+				material_builder.add(ptype, std::move(uptr));
```
Comment (hotstreams) [inline src/material_system/material.cpp:241]: a che ne material_builder.add(ptype,deserializeUniform(reader));

nah lishniy move

File: src/material_system/material.cpp [hunk 13]
```diff
@@ -213,6 +218,42 @@ Material::Material() noexcept
...
+			}
+		}
+
+		return material_builder.build();
```
Comment (hotstreams) [inline src/material_system/material.cpp:245]: a che esly material - custom ??????????////////

File: src/material_system/material_builder.cpp [hunk 14]
```diff
@@ -192,3 +192,8 @@ MaterialType MaterialBuilder::getMaterialType() const noexcept {
 
     return MaterialType{std::move(props), material->shading};
 }
+
+MaterialBuilder& MaterialBuilder::add(PropertyType type, std::unique_ptr<Uniform> uniform) {
```
Comment (hotstreams) [inline src/material_system/material_builder.cpp:196]: rip

**Review by hotstreams — COMMENTED** [review]
fix fast&furious

File: include/core/uniform.hpp [hunk 15]
```diff
@@ -5,21 +5,40 @@
...
+#include <texture_loader.hpp>
 
 namespace GraphicsEngine {
     class Texture;
     class ShaderProgram;
+    class Uniform;
 
     enum class UniformType { Value, Sampler, Time };
-    enum class UniformValueType { Float, Int, Uint, Vec2, Vec3, Vec4, Mat3, Mat4 };
+    enum class UniformValueType { Invalid, Float, Int, Uint, Vec2, Vec3, Vec4, Mat3, Mat4 };
```
Comment (hotstreams) [inline include/core/uniform.hpp:22]: want valid state

File: include/core/uniform.hpp [hunk 16]
```diff
@@ -5,21 +5,40 @@
...
         std::string name;
         UniformType type;
         UniformValueType value_type;
         bool changed;
+
+        virtual ByteWriter& serialize(ByteWriter& writer) const {
+        	return writer << name << type << value_type;
+        }
     public:
+    	explicit Uniform(ByteReader& reader) : changed{false} {
```
Comment (hotstreams) [inline include/core/uniform.hpp:38]: dont want

File: include/core/uniform.hpp [hunk 17]
```diff
@@ -30,16 +49,25 @@ namespace GraphicsEngine {
...
         T value;
 
-        constexpr UniformValueType getUniformValueType();
+        [[nodiscard]] constexpr UniformValueType getUniformValueType() const;
         UniformValue(const std::string& name, UniformType type, const T& value);
+
+        ByteWriter& serializeFields(ByteWriter& writer) const;
+        ByteWriter& serialize(ByteWriter& writer) const override;
     public:
+    	explicit UniformValue(ByteReader& reader);
```
Comment (hotstreams) [inline include/core/uniform.hpp:69]: tresh

File: include/core/uniform.hpp [hunk 18]
```diff
@@ -68,7 +100,15 @@ namespace GraphicsEngine {
     class UniformTime : public UniformValue<float> {
     private:
         std::chrono::time_point<std::chrono::steady_clock> start;
+    protected:
+	    ByteWriter& serialize(ByteWriter& writer) const override {
+	    	writer << UniformType::Time;
```
Comment (hotstreams) [inline include/core/uniform.hpp:105]: hmmmmm

File: include/material_system/material.hpp (hunk 7 above)
Comment (Tehsapper) [reply include/material_system/material.hpp:71]: a nah mu4it jopy esli 1) serializer nujen polniy dostup v predstavlenie classa
esli sdelat friend to ne 04 + pridetsya dlya kajdoy template instantiation pisat
2) objecty poh *kak* ego serializiruyut. emy ne poh tolko *4to* serializiruyut

tipa mogy soglasitsya na templated serialize + deserialize methody vnutri cla𐌔𐌔a (i to pridetsya delat tresh default constructor ibo ty j ne ho4ew constructor iz readera)
no otdelniy interface dlya etogo eto perebor

File: src/material_system/material.cpp (hunk 13 above)
Comment (Tehsapper) [reply src/material_system/material.cpp:245]: custom material eto drugaya cool story

File: include/core/uniform.hpp (hunk 15 above)
Comment (Tehsapper) [reply include/core/uniform.hpp:22]: soryan need poison value

**Pull request closed without merging**